Plugin Compatibility with or with out API fix

There was a issue created in the API for email. The plugin seems like is not compatible and need to ensure it works if there is a fix applied or not the environment.

Current dataops plugin version on last version of saagie : 

saagies-mbp:Gradle shivakumar$ gradle -b build.gradle.projectsCreateJob projectsCreateJob
Task :projectsCreateJob FAILED
Something went wrong when creating project job: {“errors”:[{“message”:“Job not valid”,“extensions”:{“emails”:“not empty”,“classification”:“ValidationError”}}],“data”:null}
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ‘:projectsCreateJob’.
Something went wrong when creating project job: {“errors”:[{“message”:“Job not valid”,“extensions”:{“emails”:“not empty”,“classification”:“ValidationErr


Looks like we are sending the alerting part when creating a job even if nothing is set on the gradle file part.

--- a/src/main/groovy/io/saagie/plugin/dataops/models/Alerting.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/Alerting.groovy
@@ -2,15 +2,15 @@ package io.saagie.plugin.dataops.models
 
 class Alerting implements IMapable {
     List<String> emails = []
+    List<String> logins = []
     List<String> statusList = []
 
     @Override
     Map toMap() {
-        if (emails.empty || statusList.empty) {
-            return null
-        }
+        if (emails.empty || logins.empty || statusList.empty) return null
         return [
-            emails: emails,
+            emails    : emails,
+            logins    : logins,
             statusList: statusList
         ]
     }

--- a/src/main/groovy/io/saagie/plugin/dataops/models/Alerting.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/Alerting.groovy
@@ -1,6 +1,17 @@
 package io.saagie.plugin.dataops.models
 
-class Alerting {
+class Alerting implements IMapable {
     List<String> emails = []
     List<String> statusList = []
+
+    @Override
+    Map toMap() {
+        if (emails.empty || statusList.empty) {
+            return null
+        }
+        return [
+            emails: emails,
+            statusList: statusList
+        ]
+    }
 }

--- a/src/main/groovy/io/saagie/plugin/dataops/models/DockerInfos.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/DockerInfos.groovy
@@ -9,8 +9,8 @@ class DockerInfos implements IMapable {
     Map toMap() {
         if (image && login && password) {
             return [
-                image: image,
-                login: login,
+                image   : image,
+                login   : login,
                 password: password,
             ]
         }

--- a/src/main/groovy/io/saagie/plugin/dataops/models/DockerInfos.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/DockerInfos.groovy
@@ -1,7 +1,19 @@
 package io.saagie.plugin.dataops.models
 
-class DockerInfos {
-    String image = ''
-    String login = ''
-    String password = ''
+class DockerInfos implements IMapable {
+    String image
+    String login
+    String password
+
+    @Override
+    Map toMap() {
+        if (image && login && password) {
+            return [
+                image: image,
+                login: login,
+                password: password,
+            ]
+        }
+        return null
+    }
 }

--- a/src/main/groovy/io/saagie/plugin/dataops/models/ExposedPort.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/ExposedPort.groovy
@@ -1,0 +1,21 @@
+package io.saagie.plugin.dataops.models
+
+class ExposedPort implements IMapable {
+    String name
+    Integer port
+    Boolean isRewriteUrl
+    String basePathVariableName
+    Boolean isAuthenticationRequired
+
+    @Override
+    Map toMap() {
+        if (!port) return null
+        return [
+            name                    : name,
+            port                    : port,
+            isRewriteUrl            : isRewriteUrl,
+            basePathVariableName    : basePathVariableName,
+            isAuthenticationRequired: isAuthenticationRequired,
+        ]
+    }
+}

--- a/src/main/groovy/io/saagie/plugin/dataops/models/IMapable.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/IMapable.groovy
@@ -1,0 +1,5 @@
+package io.saagie.plugin.dataops.models
+
+interface IMapable {
+    Map toMap()
+}

--- a/src/main/groovy/io/saagie/plugin/dataops/models/Job.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/Job.groovy
@@ -1,6 +1,6 @@
 package io.saagie.plugin.dataops.models
 
-class Job {
+class Job implements IMapable {
     String name
     String id
     String projectId
@@ -14,5 +14,24 @@ class Job {
 
     Object alerting(Closure closure) {
         alerting.with(closure)
+    }
+
+    @Override
+    Map toMap() {
+        if (name && id && projectId && description && category && technology && cronScheduling) {
+            return [
+                name: name,
+                id: id,
+                projectId: projectId,
+                description: description,
+                category: category,
+                technology: technology,
+                cronScheduling: cronScheduling,
+                isScheduled: isScheduled,
+                isStreaming: isStreaming,
+                alerting: alerting.toMap(),
+            ]
+        }
+        return null
     }
 }

--- a/src/main/groovy/io/saagie/plugin/dataops/models/Job.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/Job.groovy
@@ -18,20 +18,17 @@ class Job implements IMapable {
 
     @Override
     Map toMap() {
-        if (name && id && projectId && description && category && technology && cronScheduling) {
-            return [
-                name: name,
-                id: id,
-                projectId: projectId,
-                description: description,
-                category: category,
-                technology: technology,
-                cronScheduling: cronScheduling,
-                isScheduled: isScheduled,
-                isStreaming: isStreaming,
-                alerting: alerting.toMap(),
-            ]
-        }
-        return null
+        return [
+            name          : name,
+            id            : id,
+            projectId     : projectId,
+            description   : description,
+            category      : category,
+            technology    : [id: technology],
+            cronScheduling: cronScheduling,
+            isScheduled   : isScheduled,
+            isStreaming   : isStreaming,
+            alerting      : alerting.toMap(),
+        ]
     }
 }

--- a/src/main/groovy/io/saagie/plugin/dataops/models/JobInstance.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/JobInstance.groovy
@@ -1,5 +1,13 @@
 package io.saagie.plugin.dataops.models
 
-class JobInstance {
+class JobInstance implements IMapable {
     String id
+
+    @Override
+    Map toMap() {
+        if (id) {
+            return [id: id]
+        }
+        return null
+    }
 }

--- a/src/main/groovy/io/saagie/plugin/dataops/models/JobVersion.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/JobVersion.groovy
@@ -1,6 +1,6 @@
 package io.saagie.plugin.dataops.models
 
-class JobVersion {
+class JobVersion implements IMapable {
     String commandLine
     String releaseNote
     String runtimeVersion
@@ -18,5 +18,22 @@ class JobVersion {
 
     Object resources(Closure closure) {
         resources.with(closure)
+    }
+
+    @Override
+    Map toMap() {
+        if ((commandLine && releaseNote && runtimeVersion) ||
+            packageInfo.toMap() || dockerInfos.toMap() || resources.toMap()
+        ) {
+            return [
+                commandLine: commandLine,
+                releaseNote: releaseNote,
+                runtimeVersion: runtimeVersion,
+                packageInfo: packageInfo.toMap(),
+                dockerInfos: dockerInfos.toMap(),
+                resources: resources.toMap(),
+            ]
+        }
+        return null;
     }
 }

--- a/src/main/groovy/io/saagie/plugin/dataops/models/JobVersion.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/JobVersion.groovy
@@ -4,6 +4,11 @@ class JobVersion implements IMapable {
     String commandLine
     String releaseNote
     String runtimeVersion
+    Boolean usePreviousArtifact = false
+
+    List<String> volume = []
+    List<ExposedPort> exposedPorts = []
+
     PackageInfo packageInfo = new PackageInfo()
     DockerInfos dockerInfos = new DockerInfos()
     Resources resources = new Resources()
@@ -22,18 +27,16 @@ class JobVersion implements IMapable {
 
     @Override
     Map toMap() {
-        if ((commandLine && releaseNote && runtimeVersion) ||
-            packageInfo.toMap() || dockerInfos.toMap() || resources.toMap()
-        ) {
-            return [
-                commandLine: commandLine,
-                releaseNote: releaseNote,
-                runtimeVersion: runtimeVersion,
-                packageInfo: packageInfo.toMap(),
-                dockerInfos: dockerInfos.toMap(),
-                resources: resources.toMap(),
-            ]
-        }
-        return null;
+        return [
+            commandLine        : commandLine,
+            releaseNote        : releaseNote,
+            runtimeVersion     : runtimeVersion,
+            volume             : volume,
+            usePreviousArtifact: usePreviousArtifact,
+            exposedPorts       : exposedPorts,
+            packageInfo        : packageInfo.toMap(),
+            dockerInfos        : dockerInfos.toMap(),
+            resources          : resources.toMap(),
+        ]
     }
 }

--- a/src/main/groovy/io/saagie/plugin/dataops/models/PackageInfo.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/PackageInfo.groovy
@@ -1,5 +1,13 @@
 package io.saagie.plugin.dataops.models
 
-class PackageInfo {
+class PackageInfo implements IMapable {
     String name
+
+    @Override
+    Map toMap() {
+        if (name) {
+            return [name: name]
+        }
+        return null
+    }
 }

--- a/src/main/groovy/io/saagie/plugin/dataops/models/Pipeline.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/Pipeline.groovy
@@ -1,11 +1,11 @@
 package io.saagie.plugin.dataops.models
 
 class Pipeline implements IMapable {
-    String id = null
+    String id
     String name
-    String description = null
+    String description
     Boolean isScheduled = false
-    String cronScheduling = null
+    String cronScheduling
     Alerting alerting = new Alerting()
 
     Object alerting(Closure closure) {
@@ -16,12 +16,12 @@ class Pipeline implements IMapable {
     Map toMap() {
         if (name) {
             return [
-                id: id,
-                name: name,
-                description: description,
-                isScheduled: isScheduled,
+                id            : id,
+                name          : name,
+                description   : description,
+                isScheduled   : isScheduled,
                 cronScheduling: cronScheduling,
-                alerting: alerting.toMap(),
+                alerting      : alerting.toMap(),
             ]
         }
         return null

--- a/src/main/groovy/io/saagie/plugin/dataops/models/Pipeline.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/Pipeline.groovy
@@ -1,6 +1,6 @@
 package io.saagie.plugin.dataops.models
 
-class Pipeline {
+class Pipeline implements IMapable {
     String id = null
     String name
     String description = null
@@ -10,5 +10,20 @@ class Pipeline {
 
     Object alerting(Closure closure) {
         alerting.with(closure)
+    }
+
+    @Override
+    Map toMap() {
+        if (name) {
+            return [
+                id: id,
+                name: name,
+                description: description,
+                isScheduled: isScheduled,
+                cronScheduling: cronScheduling,
+                alerting: alerting.toMap(),
+            ]
+        }
+        return null
     }
 }

--- a/src/main/groovy/io/saagie/plugin/dataops/models/PipelineInstance.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/PipelineInstance.groovy
@@ -1,5 +1,11 @@
 package io.saagie.plugin.dataops.models
 
-class PipelineInstance {
+class PipelineInstance implements IMapable {
     String id
+
+    @Override
+    Map toMap() {
+        if (id) return [id: id]
+        return null
+    }
 }

--- a/src/main/groovy/io/saagie/plugin/dataops/models/PipelineVersion.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/PipelineVersion.groovy
@@ -9,7 +9,7 @@ class PipelineVersion implements IMapable {
         if (releaseNote && !jobs.empty) {
             return [
                 releaseNote: releaseNote,
-                jobs: jobs
+                jobs       : jobs
             ]
         }
         return null

--- a/src/main/groovy/io/saagie/plugin/dataops/models/PipelineVersion.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/PipelineVersion.groovy
@@ -1,6 +1,17 @@
 package io.saagie.plugin.dataops.models
 
-class PipelineVersion {
+class PipelineVersion implements IMapable {
     String releaseNote
     List<String> jobs = []
+
+    @Override
+    Map toMap() {
+        if (releaseNote && !jobs.empty) {
+            return [
+                releaseNote: releaseNote,
+                jobs: jobs
+            ]
+        }
+        return null
+    }
 }

--- a/src/main/groovy/io/saagie/plugin/dataops/models/Project.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/Project.groovy
@@ -12,12 +12,12 @@ class Project implements IMapable {
     Map toMap() {
         if (id && name && creator && description && jobsCount && status) {
             return [
-                id: id,
-                name: name,
-                creator: creator,
+                id         : id,
+                name       : name,
+                creator    : creator,
                 description: description,
-                jobsCount: jobsCount,
-                status: status
+                jobsCount  : jobsCount,
+                status     : status
             ]
         }
         return null

--- a/src/main/groovy/io/saagie/plugin/dataops/models/Project.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/Project.groovy
@@ -1,10 +1,25 @@
 package io.saagie.plugin.dataops.models
 
-class Project {
+class Project implements IMapable {
     String id
     String name
     String creator
     String description
     Integer jobsCount
     String status
+
+    @Override
+    Map toMap() {
+        if (id && name && creator && description && jobsCount && status) {
+            return [
+                id: id,
+                name: name,
+                creator: creator,
+                description: description,
+                jobsCount: jobsCount,
+                status: status
+            ]
+        }
+        return null
+    }
 }

--- a/src/main/groovy/io/saagie/plugin/dataops/models/Resources.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/Resources.groovy
@@ -8,8 +8,8 @@ class Resources implements IMapable {
     Map toMap() {
         if (cpu && disk && memory) {
             return [
-                cpu: cpu,
-                disk: disk,
+                cpu   : cpu,
+                disk  : disk,
                 memory: memory,
             ]
         }

--- a/src/main/groovy/io/saagie/plugin/dataops/models/Resources.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/models/Resources.groovy
@@ -1,7 +1,18 @@
 package io.saagie.plugin.dataops.models
 
-class Resources {
+class Resources implements IMapable {
     Float cpu
     Float disk
     Float memory
+
+    Map toMap() {
+        if (cpu && disk && memory) {
+            return [
+                cpu: cpu,
+                disk: disk,
+                memory: memory,
+            ]
+        }
+        return null
+    }
 }

--- a/src/main/groovy/io/saagie/plugin/dataops/utils/SaagieUtils.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/utils/SaagieUtils.groovy
@@ -22,7 +22,7 @@ import org.gradle.api.logging.Logging
 @TypeChecked
 class SaagieUtils {
     static final Logger logger = Logging.getLogger(SaagieUtils.class)
-    static final MediaType JSON = MediaType.parse 'application/json; charset=utf-8'
+    static final MediaType JSON = MediaType.parse('application/json; charset=utf-8')
     DataOpsExtension configuration
 
     SaagieUtils(DataOpsExtension configuration) {
@@ -162,7 +162,6 @@ class SaagieUtils {
         def jsonGenerator = new JsonGenerator.Options()
             .excludeNulls()
             .excludeFieldsByName('dockerInfos') // TODO: remove this line when `dockerInfos` will be available
-            .addConverter(String) { String value, String key -> key == 'technology' ? [id: value] : value }
             .addConverter(JobVersion) { JobVersion value ->
                 value.packageInfo.name = file.name
                 return value
@@ -170,7 +169,7 @@ class SaagieUtils {
             .build()
 
         def gqVariables = jsonGenerator.toJson([
-            job: job.toMap(),
+            job       : job.toMap(),
             jobVersion: jobVersion.toMap()
         ])
 
@@ -197,7 +196,7 @@ class SaagieUtils {
             .build()
 
         def gqVariables = jsonGenerator.toJson([
-            job: job,
+            job: job.toMap(),
         ])
 
         def updateProjectJob = gq('''
@@ -221,7 +220,6 @@ class SaagieUtils {
         def jsonGenerator = new JsonGenerator.Options()
             .excludeNulls()
             .excludeFieldsByName('dockerInfos') // TODO: remove this line when `dockerInfos` will be available
-            .addConverter(String) { String value, String key -> key == 'technology' ? [id: value] : value }
             .addConverter(JobVersion) { JobVersion value ->
                 value.packageInfo.name = file.name
                 return value
@@ -229,8 +227,8 @@ class SaagieUtils {
             .build()
 
         def gqVariables = jsonGenerator.toJson([
-            jobId: job.id,
-            jobVersion: jobVersion
+            jobId     : job.id,
+            jobVersion: jobVersion.toMap()
         ])
 
         // quick hack needed because the toJson seems to update the converted object, even with a clone
@@ -386,7 +384,7 @@ class SaagieUtils {
             .build()
 
         def gqVariables = jsonGenerator.toJson([
-            pipeline: pipeline
+            pipeline: pipeline.toMap()
         ])
 
         def editPipeline = gq('''

--- a/src/main/groovy/io/saagie/plugin/dataops/utils/SaagieUtils.groovy
+++ b/src/main/groovy/io/saagie/plugin/dataops/utils/SaagieUtils.groovy
@@ -170,8 +170,8 @@ class SaagieUtils {
             .build()
 
         def gqVariables = jsonGenerator.toJson([
-            job: job,
-            jobVersion: jobVersion
+            job: job.toMap(),
+            jobVersion: jobVersion.toMap()
         ])
 
         // quick hack needed because the toJson seems to update the converted object, even with a clone
@@ -583,5 +583,18 @@ class SaagieUtils {
             .collectEntries { field ->
                 [field.name, obj["$field.name"]]
             }
+    }
+
+    // From stackoverflow: https://stackoverflow.com/a/14754409/12188726
+    def denull(obj) {
+        if(obj instanceof Map) {
+            obj.collectEntries {k, v ->
+                if(v) [(k): denull(v)] else [:]
+            }
+        } else if(obj instanceof List) {
+            obj.collect { denull(it) }.findAll { it != null }
+        } else {
+            obj
+        }
     }
 }

--- a/src/test/groovy/io/saagie/plugin/job/JobCreateTaskTests.groovy
+++ b/src/test/groovy/io/saagie/plugin/job/JobCreateTaskTests.groovy
@@ -182,7 +182,7 @@ class JobCreateTaskTests extends Specification {
         """
 
         when:
-        BuildResult result = gradle 'projectsCreateJob'
+        BuildResult result = gradle ('projectsCreateJob', '-d')
 
         then:
         Exception e = thrown()

--- a/src/test/groovy/io/saagie/plugin/project/ProjectListTaskTests.groovy
+++ b/src/test/groovy/io/saagie/plugin/project/ProjectListTaskTests.groovy
@@ -59,9 +59,7 @@ class ProjectListTaskTests extends Specification {
         given:
         def mockedResponse = new MockResponse()
         mockedResponse.responseCode = 200
-        mockedResponse.body = """
-            {"data":{"projects":[{"id":"8321e13c-892a-4481-8552-dekzdjeijzd","name":"Test new Project"},{"id":"7f5e0374-0c45-45a3-a2f3-dkjezoijdizd","name":"Test Spark config"},{"id":"bba3511b-7b7f-44f0-9f69-djeizjdoijzj","name":"For tests"},{"id":"9feae78d-1cc0-49bd-9e63-deozjiodjeiz","name":"Test simon"}]}}
-        """
+        mockedResponse.body = """{"data":{"projects":[{"id":"8321e13c-892a-4481-8552-dekzdjeijzd","name":"Test new Project"},{"id":"7f5e0374-0c45-45a3-a2f3-dkjezoijdizd","name":"Test Spark config"},{"id":"bba3511b-7b7f-44f0-9f69-djeizjdoijzj","name":"For tests"},{"id":"9feae78d-1cc0-49bd-9e63-deozjiodjeiz","name":"Test simon"}]}}"""
 
         buildFile << """
             saagie {
@@ -80,7 +78,7 @@ class ProjectListTaskTests extends Specification {
 
         then:
         !result.output.contains('"data"')
-        result.output.contains('[{"id":"8321e13c')
+        result.output.contains("""[{"id":"8321e13c-892a-4481-8552-dekzdjeijzd","name":"Test new Project"},{"id":"7f5e0374-0c45-45a3-a2f3-dkjezoijdizd","name":"Test Spark config"},{"id":"bba3511b-7b7f-44f0-9f69-djeizjdoijzj","name":"For tests"},{"id":"9feae78d-1cc0-49bd-9e63-deozjiodjeiz","name":"Test simon"}]""")
     }
 
     def "projectsList task with bad config should fail"() {

